mizar/dp/mizar/operators/droplets/droplets_operator.py

```diff
@@ -92,14 +93,56 @@ def assign_bouncer_droplet(self, bouncer):
         droplets = set(self.store.get_all_droplets())
         if len(droplets) == 0:
             return False
-        d = random.sample(droplets, 1)[0]
+        # Read cluster_gateway_host_ip from configmap
+        cluster_gateway_host_ip = get_cluster_gateway_host_ip(self.core_api)
+        subnets = self.store.get_nets_in_vpc(bouncer.vpc)
+        # remove portal hosts from the droplet set
+        cluster_gateway_droplet = ""
+        remote_deployed_subnet_ips = set()
+        for subnet in subnets.values():
+            if subnet.remote_deployed:
+                remote_deployed_subnet_ips.add(subnet.ip)
+                logger.info("A subnet ip {} for subnet {} has been added.".format( subnet.ip, subnet.name))
+
+        for dd in droplets:
+            if dd.ip == cluster_gateway_host_ip:
+                cluster_gateway_droplet = dd
+                logger.info("A droplet {} has been added as cluster gateway.".format(dd.ip))
+
+        if cluster_gateway_droplet != "":
+            droplets.remove(cluster_gateway_droplet)
+            logger.info("The cluster gateway droplet {} has been removed.".format(cluster_gateway_droplet))
+
+        if bouncer.get_nip() in remote_deployed_subnet_ips and cluster_gateway_droplet != "":
+            # for remote deployed subnets, use the cluster gateway host instead of picking a host as bouncer
+            d = cluster_gateway_droplet
+            logger.info("remote deployed subnet, using cluster gateway droplet {}".format(d.ip))
+        else:
+            d = random.sample(droplets, 1)[0]
+
         bouncer.set_droplet(d)
         return True
 
     def assign_divider_droplet(self, divider):
         droplets = set(self.store.get_all_droplets())
         if len(droplets) == 0:
             return False
+
+        # Read cluster_gateway_host_ip from configmap
+        cluster_gateway_host_ip = get_cluster_gateway_host_ip(self.core_api)
+
+        cluster_gateway_droplet = ""
+        for dd in droplets:
+            if dd.ip == cluster_gateway_host_ip:
+                cluster_gateway_droplet = dd
+                logger.info("The cluster gateway droplet {} has been added.".format(dd.ip))
+        if cluster_gateway_droplet != "":
+            droplets.remove(cluster_gateway_droplet)
+
+        # All the droplets have been removed as cluster gateway  host droplet
+        if len(droplets) == 0:
+            return False
+
```

Comment:
Please package this into a function, and move it to a file common for all fornax related functions.

Reply:
Moved